libbeat/publisher/queue/memqueue/broker.go

```diff
@@ -137,76 +161,80 @@ func NewQueue(
 	settings Settings,
 	inputQueueSize int,
 ) *broker {
-	var (
-		sz           = settings.Events
-		minEvents    = settings.FlushMinEvents
-		flushTimeout = settings.FlushTimeout
-	)
+	b := newQueue(logger, ackCallback, settings, inputQueueSize)
 
-	chanSize := AdjustInputQueueSize(inputQueueSize, sz)
+	// Start the queue workers
+	b.wg.Add(2)
+	go func() {
+		defer b.wg.Done()
+		b.runLoop.run()
+	}()
+	go func() {
+		defer b.wg.Done()
+		b.ackLoop.run()
+	}()
 
-	if minEvents < 1 {
-		minEvents = 1
-	}
-	if minEvents > 1 && flushTimeout <= 0 {
-		minEvents = 1
-		flushTimeout = 0
+	return b
+}
+
+// newQueue does most of the work of creating a queue from the given
+// parameters, but doesn't start the runLoop or ackLoop workers. This
+// lets us perform more granular / deterministic tests by controlling
+// when the workers are active.
+func newQueue(
+	logger *logp.Logger,
+	ackCallback func(eventCount int),
+	settings Settings,
+	inputQueueSize int,
+) *broker {
+	chanSize := AdjustInputQueueSize(inputQueueSize, settings.Events)
+
+	// Backwards compatibility: an old way to select synchronous queue
+	// behavior was to set "flush.min_events" to 0 or 1, in which case the
+	// timeout was disabled and the max get request was half the queue.
+	// (Otherwise, it would make sense to leave FlushTimeout unchanged here.)
+	if settings.MaxGetRequest <= 1 {
+		settings.FlushTimeout = 0
+		settings.MaxGetRequest = (settings.Events + 1) / 2
```

Comment:
Nit.  Events & MaxGetRequests are ints, so they could be negative, which could lead to some interesting results.  Do we have protections to make sure these are positive values?  or maybe switch to uint?